### PR TITLE
Fixes: #3217 - Add a new parameter for the kms_key_id to the ec2_vol module

### PR DESCRIPTION
**Pull request by nonshankus**
##### Issue Type:
- Feature Pull Request
##### Plugin Name:

cloud/amazon/ec2_vol.py
##### Summary:

This pull request fixes/implements #3217. For more details please refer to the according issue.

It is important to point out, that this fix will require at **least boto v2.39.0**.
Here is the according boto change:
https://github.com/boto/boto/commit/2d7796a625f9596cbadb7d00c0198e5ed84631ed
And the according pull request for boto:
https://github.com/boto/boto/pull/3064
##### Example:

```
    - name: Created encrypted volume with custom key
      ec2_vol:
        device_name: /dev/sdb
        encrypted: yes
        kms_key_id: '{{ kms_key.KeyMetadata.Arn }}'
        instance: '{{ ec2.instance_ids }}'
        region: '{{global_region}}'
        volume_size: 20
        volume_type: gp2
        delete_on_termination: true
      register: ec2_vol
```

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @nonshankus for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

**Comment by nonshankus** [comment]
Thanks @gregdek for the heads-up. I sort of expected that. :)

File: cloud/amazon/ec2_vol.py [hunk 1]
```diff
@@ -62,6 +62,11 @@
       - Enable encryption at rest for this volume.
     default: false
     version_added: "1.8"
+  kms_key_id:
+    description:
+      - Specify the id of the KMS key to use.
+    default: null
```
Comment (juliedavila) [inline cloud/amazon/ec2_vol.py:68]: `default: null` can be omitted

File: cloud/amazon/ec2_vol.py [hunk 2]
```diff
@@ -332,7 +338,10 @@ def create_volume(module, ec2, zone):
     if volume is None:
         try:
             if boto_supports_volume_encryption():
-                volume = ec2.create_volume(volume_size, zone, snapshot, volume_type, iops, encrypted)
+                if kms_key_id:
```
Comment (juliedavila) [inline cloud/amazon/ec2_vol.py:350]: Can you add a check for boto version here? Something like 
Add a new import
`from distutils.version import StrictVersion`

``` python
if StrictVersion(boto.__version__) < StrictVersion('2.39.0'):
    module.fail_json(msg="A newer version of boto is required, %s detected but 2.39.0 is required for kms_key_id" % boto.__version__)
```

File: cloud/amazon/ec2_vol.py (hunk 1 above)
Comment (nonshankus) [inline cloud/amazon/ec2_vol.py:68]: Thanks @defionscode for the review. Committed an according change to remove that particular line.

File: cloud/amazon/ec2_vol.py (hunk 2 above)
Comment (nonshankus) [inline cloud/amazon/ec2_vol.py:350]: Thanks @defionscode for the feedback. That is indeed a good point. But is that consistent behaviour? On the one hand let the module fail, when a `kms_key_id` is given, and on the other gracefully ignore the `encryption` attribute, when the boto version is too old for it? Wouldn't it be more consistent to also ignore it?

Comment (juliedavila) [inline cloud/amazon/ec2_vol.py:350]: For most things I'd prob be ok with what you're saying but the last thing I would want is for a user to think that a kms key was being used for encryption, as a result of a non-error during executing, when in fact it isnt.

**Comment by kalpatel01** [comment]
https://github.com/ansible/ansible-modules-core/pull/3064 Have something already for this.  It does have one extra check.

**Comment by juliedavila** [comment]
@kalpatel01 ah! thanks for pointing that out. I'd like for your PR to take priority as yours was first. I'll review it soon

File: cloud/amazon/ec2_vol.py (hunk 1 above)
Comment (nonshankus) [inline cloud/amazon/ec2_vol.py]: Done with 71cbec0f2722ff11449c604e71e18293e7fd1e74.

Comment (nonshankus) [inline cloud/amazon/ec2_vol.py:68]: Accidentally removed the wrong line. What confused me was, that other parameters also have this line (`default: null`).

File: cloud/amazon/ec2_vol.py (hunk 2 above)
Comment (nonshankus) [inline cloud/amazon/ec2_vol.py:350]: Point taken. Added the according check, actually merged the check from https://github.com/ansible/ansible-modules-core/pull/3064 and corrected the naming of the method.

**Comment by nonshankus** [comment]
@defionscode: As previously mentioned, merged the check from https://github.com/ansible/ansible-modules-core/pull/3064 while correcting the method naming. So this pull request here only implements the feature requested in the feature request #3217, whereas the pull request by @kalpatel01 also implements adding tags to volumes.

As the pull request of @kalpatel01 was way before mine I am perfectly fine with accepting it instead of mine.

Cheers

**Pull request closed without merging**